src/eviction/lru.c

```diff
@@ -454,71 +598,78 @@ bool evp_lru_can_evict(ocf_cache_t cache)
 	return true;
 }
 
-static bool dirty_pages_present(ocf_cache_t cache, struct ocf_user_part *part)
-{
-	uint32_t i;
-
-	for (i = 0; i < OCF_NUM_EVICTION_LISTS; i++) {
-		if (evp_lru_get_list(part, i, false)->tail != end_marker)
-			return true;
-	}
-
-	return false;
-}
-
 /* the caller must hold the metadata lock */
-uint32_t evp_lru_req_clines(ocf_cache_t cache, ocf_queue_t io_queue,
+uint32_t evp_lru_req_clines(struct ocf_request *req,
 		struct ocf_user_part *part, uint32_t cline_no)
 {
 	struct ocf_lru_iter iter;
 	uint32_t i;
 	ocf_cache_line_t cline;
+	uint64_t core_line;
+	ocf_core_id_t core_id;
+	ocf_cache_t cache = req->cache;
+	bool cl_write_lock =
+		(req->engine_cbs->get_lock_type(req) ==	ocf_engine_lock_write);
+	unsigned evp;
+	unsigned req_idx = 0;
 
 	if (cline_no == 0)
 		return 0;
 
-	lru_iter_init(&iter, cache, part, part->next_eviction_list, true);
+	evp = req->io_queue->eviction_idx++ % OCF_NUM_EVICTION_LISTS;
+
+	lru_iter_eviction_init(&iter, cache, part, evp, cl_write_lock, req);
 
 	i = 0;
 	while (i < cline_no) {
-		cline = lru_iter_next(&iter);
-
-		if (cline == end_marker)
-			break;
-
 		if (!evp_lru_can_evict(cache))
 			break;
 
-		/* Prevent evicting already locked items */
-		if (ocf_cache_line_is_used(cache, cline))
-			continue;
+		cline = lru_iter_eviction_next(&iter, &core_id, &core_line);
+
+		if (cline == end_marker)
+			break;
 
 		ENV_BUG_ON(metadata_test_dirty(cache, cline));
 
-		if (ocf_volume_is_atomic(&cache->device->volume)) {
-			/* atomic cache, we have to trim cache lines before
-			 * eviction
-			 */
-			evp_lru_zero_line(cache, io_queue, cline);
-			continue;
+		/* TODO: if atomic mode is restored, need to zero metadata
+		 * before proceeding with cleaning (see version <= 20.12) */
+
+		/* find next unmapped cacheline in request */
+		while (req_idx + 1 < req->core_line_count &&
+				req->map[req_idx].status != LOOKUP_MISS) {
+			req_idx++;
 		}
 
+		ENV_BUG_ON(req->map[req_idx].status != LOOKUP_MISS);
```

Comment:
If the last core line from request would have status different than `LOOKUP_MISS`, `BUG_ON()` will be triggered

Reply:
that's right. this would mean that the caller requested to evict more cachelines than there are unmapped cachelines within the request.